nixos/modules/services/networking/hostapd.nix

```diff
@@ -1,222 +1,1007 @@
+            assertion = ifcfg.authentication.wpaPassword != null -> (stringLength ifcfg.authentication.wpaPassword >= 8 && stringLength ifcfg.authentication.wpaPassword <= 63);
+            message = ''hostapd interface ${interface} uses a wpaPassword of invalid length (must be in [8,63]).'';
+          }
+          {
+            assertion = ifcfg.authentication.saePasswords == [] || ifcfg.authentication.saePasswordsFile == null;
+            message = ''hostapd interface ${interface} must use only one SAE password option (saePasswords or saePasswordsFile)'';
+          }
+          {
+            assertion = ifcfg.authentication.mode == "wpa3-sae" -> (ifcfg.authentication.saePasswords != [] || ifcfg.authentication.saePasswordsFile != null);
+            message = ''hostapd interface ${interface} uses WPA3-SAE which requires defining a sae password option'';
+          }
+          {
+            assertion = ifcfg.authentication.mode == "wpa3-sae-transition" -> (ifcfg.authentication.saePasswords != [] || ifcfg.authentication.saePasswordsFile != null) && countWpaPasswordDefinitions == 1;
+            message = ''hostapd interface ${interface} uses WPA3-SAE in transition mode requires defining both a wpa password option and a sae password option'';
+          }
+          {
+            assertion = ifcfg.authentication.mode == "wpa2-sha256" -> countWpaPasswordDefinitions == 1;
+            message = ''hostapd interface ${interface} uses WPA2-SHA256 which requires defining a wpa password option'';
+          }
+        ])
+        cfg.interfaces));
+
+    environment.systemPackages = [pkgs.hostapd];
+
+    services.udev.packages = optionals (any (i: i.countryCode != null) (attrValues cfg.interfaces)) [pkgs.crda];
+
+    systemd.services.hostapd = {
+      description = "Hostapd IEEE 802.11 AP Daemon";
+
+      path = [pkgs.hostapd];
+      after = mapAttrsToList (interface: _: "sys-subsystem-net-devices-${utils.escapeSystemdPath interface}.device") cfg.interfaces;
+      bindsTo = mapAttrsToList (interface: _: "sys-subsystem-net-devices-${utils.escapeSystemdPath interface}.device") cfg.interfaces;
+      requiredBy = mapAttrsToList (interface: _: "network-link-${interface}.service") cfg.interfaces;
+      wantedBy = ["multi-user.target"];
+
+      # Create merged configuration and acl files for each interface prior to starting
+      preStart = concatStringsSep "\n" (mapAttrsToList makeInterfaceRuntimeFiles cfg.interfaces);
+
+      serviceConfig = {
+        ExecStart = "${pkgs.hostapd}/bin/hostapd ${concatStringsSep " " runtimeConfigFiles}";
+        Restart = "always";
+        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+        RuntimeDirectory = "hostapd";
+
+        # Hardening
+        LockPersonality = true;
+        MemoryDenyWriteExecute = true;
+        DevicePolicy = "closed";
+        DeviceAllow = "/dev/rfkill rw";
+        NoNewPrivileges = true;
+        PrivateUsers = false; # hostapd requires true root access.
```

Comment:
I imagine this to be a pain to debug since you don't get any meaningful errors if some capability is missing

Reply:
Yeah, I'd postpone this for now. It's non trivial (to me) to figure out exactly what hostapd could possibly want. Missing something would break it, so better change it when we are sure about what it requires.